pallets/parachain-staking/src/lib.rs

```diff
@@ -223,6 +224,857 @@ pub mod pallet {
 		pub state: CollatorStatus,
 	}
 
+	#[derive(Clone, Default, Encode, Decode, RuntimeDebug, TypeInfo)]
+	/// Type for top and bottom delegation storage item
+	pub struct Delegations<AccountId, Balance> {
```

Comment:
I find this struct (or its `impl`) a bit confusing. It seems to be either a top list or a bottom list depending on how it's called. Maybe making separate types for the two would help.

Reply:
They are already different storage items with different names `TopDelegations` and `BottomDelegations`. That should alleviate your concern.

Reply:
Initially I didn't realize that we are getting rid of the reverse-sorted bottom delegations (right?) which makes sense, I like that change. Given that, having different subclasses seems unnecessary.

My main concern with the design is that, as a "layer of abstraction," this partially bleeds the details of the pallet itself.

Practically speaking, this struct could be used inconsistently by its callers as both a bottom and top. `insert_sorted_greatest_to_least` makes no assumptions about capacity (it's up to the caller to do that in that case), yet 'top_capacity' and 'bottom_capacity' both make assumptions about capacity. This is especially tricky because both of the capacity functions test equality ('==' rather than '>='). So if we ever accidentally go past our bounds, the struct will continually tell us that we are `CapacityStatus::Partial`, which would probably trigger the caller to keep inserting more delegations.

In terms of making this struct "difficult to use incorrectly," I think we could simply add the capacity itself to the struct. This would have some advantages:

* `top_capacity` and `bottom_capacity` collapse into one fn
* this capacity fn can convert to `>=` (that could happen regardless, I suppose)
* `insert_sorted_greatest_to_least` can do proper bounds checking
* unit testing becomes trivial and is fully decoupled from the pallet itself
* a `resize` fn could be added (again, that could probably happen regardless).

Potentially, the capacity could be generic, although that would have implications on changing things at runtime.

Reply:
> Initially I didn't realize that we are getting rid of the reverse-sorted bottom delegations (right?) which makes sense, I like that change. Given that, having different subclasses seems unnecessary.

Yes, exactly

> Practically speaking, this struct could be used inconsistently by its callers as both a bottom and top. insert_sorted_greatest_to_least makes no assumptions about capacity (it's up to the caller to do that in that case), yet 'top_capacity' and 'bottom_capacity' both make assumptions about capacity.

Yes, but this is always enforced by the caller in the code prior to the call. Moreover, there is a capacity check in each of these functions that checks if it is full.

> In terms of making this struct "difficult to use incorrectly," I think we could simply add the capacity itself to the struct.

There is no need to do this because the `Config::MaxTopDelegationsPerCandidate` and `Config::MaxTopDelegationsPerCandidate` **are constants so there is no cost to getting them**. Inside the methods on CandidateMetadata that edit the delegations, we get these constants and check capacity to ensure correctness. Please audit those methods on `CandidateMetadata` and the use of the constants inside of them to check capacity.